File: src/main/java/pw/vodes/rimuru/file/AutoRoles.java
package pw.vodes.rimuru.file;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;

import org.javacord.api.event.message.reaction.ReactionAddEvent;
import org.javacord.api.event.message.reaction.ReactionRemoveEvent;
import org.javacord.api.listener.message.reaction.ReactionAddListener;
import org.javacord.api.listener.message.reaction.ReactionRemoveListener;

import com.fasterxml.jackson.core.JsonProcessingException;

import pw.vodes.rimuru.Main;
import pw.vodes.rimuru.file.sub.AutoRole;

public class AutoRoles {
	
	private static List<AutoRole> autoroles = new ArrayList<AutoRole>();
	
	@SuppressWarnings("unchecked")
	public static void load() {
		if(Main.getFiles().exists("autoroles.json")) {
			try {
				autoroles = (ArrayList<AutoRole>)Main.getMapper().readValue(Main.getFiles().read("autoroles.json"), Main.getMapper().getTypeFactory().constructCollectionType(ArrayList.class, AutoRole.class));
				for(var ar : autoroles) {
					try {
						var msg = Main.getServer().getChannelById(ar.channel_id).get().asServerTextChannel().get().getMessageById(ar.message_id).get();
						msg.canYouDelete();
						addListeners(ar);
					} catch (Exception e) {
						// Message doesnt exit
					}
				}
			} catch (JsonProcessingException e) {
				e.printStackTrace();
			}
		}
	}
	
	public static List<AutoRole> getAutoroles() {
		return autoroles;
	}
	
	public static void removeAutoRole(AutoRole ar) {
		try {
			var msg = Main.getServer().getTextChannelById(ar.channel_id).get().getMessageById(ar.message_id).get();

			if(ar.addListener != null) {
				msg.removeListener(ReactionAddListener.class, ar.addListener);
			}
			
			if(ar.removeListener != null) {
				msg.removeListener(ReactionRemoveListener.class, ar.removeListener);
			}
		} catch (InterruptedException | ExecutionException e) {
			e.printStackTrace();
		}

		autoroles.remove(ar);
		save();
	}
	
	private static void save() {
		try {
			Main.getFiles().write("autoroles.json", Main.getMapper().writerWithDefaultPrettyPrinter().writeValueAsString(autoroles));
		} catch (JsonProcessingException e) {
			e.printStackTrace();
		}
	}
	
	public static void addAutorole(String channel_id, String message_id, String role_id) {
		var ar = new AutoRole(channel_id, message_id, role_id);
		addListeners(ar);
		autoroles.add(ar);
		save();
	}
	
	private static void addListeners(AutoRole ar) {
		var role = Main.getServer().getRoleById(ar.role_id).get();
		
		try {
			var msg = Main.getServer().getTextChannelById(ar.channel_id).get().getMessageById(ar.message_id).get();

			var addListener = new ReactionAddListener() {
				@Override
				public void onReactionAdd(ReactionAddEvent event) {
					if (!role.hasUser(event.getUser().get())) {
						role.addUser(event.getUser().get());
					}
				}
			};

			var removeListener = new ReactionRemoveListener() {
				@Override
				public void onReactionRemove(ReactionRemoveEvent event) {
					if (role.hasUser(event.getUser().get())) {
						role.removeUser(event.getUser().get());
					}
				}
			};
			
			msg.addReactionAddListener(ar.addListener = addListener);
			msg.addReactionRemoveListener(ar.removeListener = removeListener);
		} catch (InterruptedException | ExecutionException e) {
			e.printStackTrace();
		}
	}

}

File: src/main/kotlin/pw/vodes/rimurukt/file/AutoRoles.kt
package pw.vodes.rimurukt.file

import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import kotlinx.serialization.encodeToString
import org.javacord.api.entity.channel.ServerTextChannel
import org.javacord.api.entity.message.Message
import org.javacord.api.entity.permission.Role
import org.javacord.api.entity.server.Server
import org.javacord.api.listener.message.reaction.ReactionAddListener
import org.javacord.api.listener.message.reaction.ReactionRemoveListener
import pw.vodes.rimurukt.Main
import pw.vodes.rimurukt.json
import java.io.File
import kotlin.jvm.optionals.getOrNull

@Serializable
data class AutoRole(val serverID: String, val channelID: String, val messageID: String, val roleID: String) {

    @Transient
    var reactionAddListener: ReactionAddListener? = null

    @Transient
    var reactionRemoveListener: ReactionRemoveListener? = null

    private fun server(): Server? {
        return Main.api.getServerById(this.serverID).getOrNull()
    }

    private fun channel(): ServerTextChannel? {
        return server()?.getTextChannelById(channelID)?.getOrNull()
    }

    fun role(): Role? {
        return server()?.getRoleById(roleID)?.getOrNull()
    }

    fun message(): Message? {
        return channel()?.getMessageById(messageID)?.get()
    }
}


object AutoRoles {
    private val file = File(Main.appDir, "autoroles.json")
    private var autoroles = mutableListOf<AutoRole>()

    fun load() {
        if (file.exists())
            autoroles = json.decodeFromString(file.readText())

        autoroles.forEach {
            try {
                val message = it.message()!!
                message.canYouDelete()
                addListeners(it)
            } catch (_: Exception) {
            }
        }
    }

    private fun save() {
        file.writeText(json.encodeToString(autoroles))
    }

    fun addAutoRole(serverID: String, channelID: String, messageID: String, roleID: String): Boolean {
        val ar = AutoRole(serverID, channelID, messageID, roleID)
        val result = addListeners(ar)
        if (!result)
            return false

        autoroles.add(ar)
        save()
        return true
    }

    fun removeAutoRole(autoRole: AutoRole) {
        autoroles.remove(autoRole)
        save()

        val message = autoRole.message() ?: return

        if (autoRole.reactionAddListener != null)
            message.removeListener(ReactionAddListener::class.java, autoRole.reactionAddListener)

        if (autoRole.reactionRemoveListener != null)
            message.removeListener(ReactionRemoveListener::class.java, autoRole.reactionRemoveListener)
    }

    private fun addListeners(autoRole: AutoRole): Boolean {
        val role = autoRole.role() ?: return false
        try {
            val message = autoRole.message() ?: return false

            message.addReactionAddListener(ReactionAddListener { event ->
                val user = event.user.getOrNull() ?: return@ReactionAddListener
                if (role.hasUser(user))
                    role.addUser(user)
            }.also { autoRole.reactionAddListener = it })

            message.addReactionRemoveListener(ReactionRemoveListener { event ->
                val user = event.user.getOrNull() ?: return@ReactionRemoveListener
                if (role.hasUser(user))
                    role.removeUser(user)
            }.also { autoRole.reactionRemoveListener = it })

            return true
        } catch (_: Exception) {
        }
        return false
    }
}